speaking-portal-project/src/animationFactory/animationProcessor.ts

```diff
@@ -0,0 +1,155 @@
+import { MouthCue, MouthCueArray } from '../types'
+import fs from 'fs'
+
+export async function mouthCuesToInputFile({
+    characterSelect,
+    mouthCues,
+    outputPath = './tmp/input.txt',
+}: {
+    characterSelect: string
+    mouthCues: MouthCue[]
+    outputPath?: string
+}) {
+    try {
+        await fs.promises.writeFile(outputPath, generateFrameData(characterSelect, mouthCues), {
+            flag: 'w',
+        })
+        /* Lines 17-20 for testing purposes only
+        const contents = await fs.promises.readFile(outputPath, 'utf-8')
+        console.log(contents);
+
+        return contents
+        */
+    } catch (err) {
+        console.log(err)
+        return Error('MouthCuesFileError')
+    }
+}
+
+/*
+This converts  data from the MouthCue type into a format for ffmpeg
+*/
+export function generateFrameData(avatar: string, mouthCues: MouthCue[]) {
+    let frameData = ''
+    let blink = false
+    let lastBlink = -1
+    let lastPoseChange = -1
+    let idle = true
+    let breathPhase = 0 // 0 = rest, 1 = inhaling, 2 = lungs full, 3 = exhaling
+
+    console.log("Adding realism...")
+    for (const mouthCuesKey in mouthCues) {
+        // Duration of frame is calculated based on the mouth cue
+        let frameDur = (mouthCues[mouthCuesKey].end - mouthCues[mouthCuesKey].start).toFixed(2)
+        let currentSec = parseInt(mouthCues[mouthCuesKey].start.toFixed(0))
+
+        // TODO: use for testing. After confirming that blinking and other realism is good, delete this
+        console.log("t=%d, frameDur: %d, lastBlink: %d, idling? %s, lastPoseChange: %d",
+                    currentSec, frameDur, lastBlink, lastPoseChange, idle, lastPoseChange)
+
+        // Start building frame's path
+        frameData.concat(`file ../images/${avatar}/${mouthCues[mouthCuesKey].value}`)
+
+        // Blinking
+        blink = willBlink(currentSec, frameDur, lastBlink)
+        if (blink) {
+            lastBlink += 1
```

Comment:
Is lastBlink just holding a variable relating to how many times it blinks? or should we be making lastBlink = the frame it last blinked?

Reply:
Last blink is the second in which the character last blinked. It is used so that the avatar doesn't blink in the same second.

If you see double blinks, it's because the avatar probably blinked at the end of a second and then again at the beginning of another second :)